Refactor docs for ESM

Falke-Design opened the pull request:
Fixes: #8352 

No need to use the bundled version anymore:
Closes: #5807 
Closes: #7615
Closes: #5696

- Replaces all `var` with `const` / `let`
- Changing all examples to `modules`
- Replacing all factories `L.xYZ()` to `new XYZ()`
- Adding `importmap` to Tutorial
- Updating Plunkr code

jonkoops commented on docs/download.md:
This page likely requires more changes, as the "Using a Hosted Version of Leaflet" still refers to loading the Leaflet script directly, instead of using an import map. Ideally, all versions would use the import map and provide instructions on where in the DOM it should be placed (the head), and  would also have a section on [`modulepreload`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/modulepreload).

jonkoops commented on src/core/Class.leafdoc:
```diff
@@ -10,19 +10,19 @@ In addition to implementing a simple classical inheritance model, it introduces
-	alert(this.greeter + ', ' + name)
-	}
+const MyClass = Class.extend({
+    initialize: function (greeter) {
```
The `function` keyword can be removed here.
```suggestion
    initialize(greeter) {
```

jonkoops commented on src/core/Class.leafdoc:
```diff
@@ -10,19 +10,19 @@ In addition to implementing a simple classical inheritance model, it introduces
+        // class constructor
+    },
+
+    greet: function (name) {
```
```suggestion
    greet(name) {
```

jonkoops commented on spec/suites/layer/tile/GridLayerSpec.js:
```diff
@@ -562,8 +562,8 @@ describe('GridLayer', () => {
-			var pending = 0;
-			for (var key in grid._tiles) {
+			let pending = 0;
+			for (let key in grid._tiles) {
```
```suggestion
			for (const key in grid._tiles) {
```

jonkoops commented on eslint.config.js:
```diff
@@ -16,6 +16,7 @@ export default [
 			'docs/examples/geojson/sample-geojson.js',
 			'docs/examples/map-panes/eu-countries.js',
 			'docs/examples/extending/extending-2-layers.md',
+			'docs/examples/quick-start/index.md',
```
Why is this one excluded?

Falke-Design replied:
Eslint is throwing an lint error for `"imports":` in the importmap. I wasn't able to ignore only this specific error for this line / file or fix it with updating eslint.
```
<script type="importmap">
	{
		"imports": {
			"leaflet": "https://unpkg.com/leaflet@{{ site.latest_leaflet_version}}/dist/leaflet.js"
		},
```

jonkoops replied:
Good enough a reason for me

jonkoops commented on docs/examples/zoom-levels/example-zero.md:
```diff
@@ -2,18 +2,22 @@
 
 	map.setView([0, 0], 0);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
Let's use `globalThis`, which is more agnostic as a global.
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/zoom-levels/example-setzoom.md:
```diff
@@ -41,4 +42,7 @@ title: Zoom Levels Tutorial
 
 	map.setView([0, 0], 0);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

Also, do we actually need to expose these globals?

jonkoops commented on docs/examples/zoom-levels/example-scale.md:
```diff
@@ -26,4 +27,7 @@ title: Zoom Levels Tutorial
 
 	map.setView([0, 0], 0);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/zoom-levels/example-fractional.md:
```diff
@@ -50,4 +51,7 @@ title: Zoom Levels Tutorial
 
 	map.setView([0, 0], 0);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/zoom-levels/example-delta.md:
```diff
@@ -35,4 +36,7 @@ title: Zoom Levels Tutorial
 
 	map.setView([0, 0], 0);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/wms/wms-example3.md:
```diff
@@ -2,33 +2,36 @@
 
 	basemaps.Topography.addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/wms/wms-example2.md:
```diff
@@ -2,15 +2,18 @@
 		layers: 'SRTM30-Colored-Hillshade'
 	}).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/quick-start/example.md:
```diff
@@ -44,4 +45,6 @@ customMapContainer: "true"
 
 	map.on('click', onMapClick);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/quick-start/example-overlays.md:
```diff
@@ -4,29 +4,31 @@ title: Quick Start
 	]).addTo(map);
 
-
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/quick-start/example-basic.md:
```diff
@@ -4,13 +4,16 @@ title: Quick Start
 		attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
 	}).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/overlays/index.md:
````diff
@@ -225,24 +225,24 @@ Alternatively, you can create the SVG element in HTML code:
 ```
 
-Although SVGOverlay does not have its own unique options, it inherits a variety of options from ImageOverlay, Interactive layer and Layer.
+Although SVGOverlay does not have its own unique options, it inherits a constiety of options from ImageOverlay, Interactive layer and Layer.
````
```suggestion
Although SVGOverlay does not have its own unique options, it inherits a variety of options from ImageOverlay, Interactive layer and Layer.
```

jonkoops commented on docs/examples/overlays/example-video.md:
```diff
@@ -56,4 +58,6 @@ title: Video Overlay Tutorial (video with controls)
 		const playControl = (new MyPlayControl()).addTo(map);
 	});
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/overlays/example-svg.md:
```diff
@@ -14,12 +16,15 @@ title: SVG Overlay Tutorial
 		interactive: true
 	}).addTo(map);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/overlays/example-nocontrols.md:
```diff
@@ -28,4 +30,7 @@ title: Video Overlay Tutorial
 	}).addTo(map);
 
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/overlays/example-image.md:
```diff
@@ -2,26 +2,31 @@
+	new Rectangle(latLngBounds).addTo(map);
 	map.fitBounds(latLngBounds);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/mobile/example.md:
```diff
@@ -35,4 +36,7 @@ css: "body {
 
 	map.locate({setView: true, maxZoom: 16});
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/map-panes/example.md:
```diff
@@ -17,21 +18,24 @@ title: Custom Icons Tutorial
 
 	map.setView({lat: 47.040182144806664, lng: 9.667968750000002}, 4);
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/layers-control/example.md:
```diff
@@ -2,48 +2,52 @@
+	layerControl.addBaseLayer(openTopoMap, 'OpenTopoMap');
+	layerControl.addOverlay(parks, 'Parks');
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/geojson/geojson.html:
This one could not be ported?

Falke-Design replied:
An old relic that doesn't work even now. I think it was generated from the `.md` file many years ago and then accidentally committed. https://leafletjs.com/examples/geojson/geojson.html

jonkoops replied:
Got it, consider this resolved.

jonkoops commented on docs/examples/geojson/geojson-example.html:
This one could not be ported?

Falke-Design replied:
An old relic that doesn't work even now. I think it was generated from the `.md` file many years ago and then accidentally committed. https://leafletjs.com/examples/geojson/geojson-example.html

jonkoops commented on docs/examples/geojson/example.md:
```diff
@@ -63,13 +65,15 @@ title: GeoJSON tutorial
 		onEachFeature
 	}).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/watermark.md:
```diff
@@ -27,11 +29,9 @@ title: Watermark control
-	const watermarkControl = L.control.watermark({position: 'bottomleft'}).addTo(map);
+	const watermarkControl = new Control.Watermark({position: 'bottomleft'}).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/tilt.md:
```diff
@@ -51,17 +52,19 @@ title: Tilt handler
 		attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
 	}).addTo(map);
 	
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/pixelorigin.md:
```diff
@@ -83,4 +83,6 @@ title: Grid coordinates
 
 	info();
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/kittenlayer.md:
```diff
@@ -2,29 +2,29 @@
 	map.addLayer(kittenTiles);
-	
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/gridcoords.md:
```diff
@@ -23,11 +24,9 @@ title: Grid coordinates
 	map.addLayer(debugCoordsGrid);
-	
+
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/class-diagram.md:
```diff
@@ -6,19 +6,22 @@ css: "#map {
 
 	map.fitBounds(bounds);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/extending/canvascircles.md:
```diff
@@ -28,11 +29,9 @@ title: CanvasCircles
+	const cavasGridLayer = new GridLayer.CanvasCircles();
 	map.addLayer(cavasGridLayer);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/custom-icons/example.md:
```diff
@@ -24,8 +25,10 @@ title: Custom Icons Tutorial
+	const mRed = new Marker([51.495, -0.083], {icon: redIcon}).bindPopup('I am a red leaf.').addTo(map);
+	const mOrange = new Marker([51.49, -0.1], {icon: orangeIcon}).bindPopup('I am an orange leaf.').addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/custom-icons/example-one-icon.md:
```diff
@@ -22,6 +23,8 @@ title: Custom Icons Tutorial
-	const mGreen = L.marker([51.5, -0.09], {icon: greenIcon}).addTo(map);
+	const mGreen = new Marker([51.5, -0.09], {icon: greenIcon}).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/crs-simple/crs-simple-example3.md:
```diff
@@ -2,37 +2,40 @@
 
 	map.setView(xy(120, 70), 1);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/crs-simple/crs-simple-example2.md:
```diff
@@ -2,19 +2,22 @@
 
 	map.setView([70, 120], 1);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/crs-simple/crs-simple-example1.md:
```diff
@@ -2,15 +2,18 @@
 
 	map.fitBounds(bounds);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/choropleth/example.md:
```diff
@@ -145,4 +146,6 @@ css: "#map {
 
 	legend.addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/choropleth/example-color.md:
```diff
@@ -36,8 +37,10 @@ title: Choropleth Tutorial
 		style,
 	}).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/choropleth/example-basic.md:
```diff
@@ -3,16 +3,19 @@ layout: tutorial_frame
-	const geojson = L.geoJson(statesData).addTo(map);
+	const geojson = new GeoJSON(statesData).addTo(map);
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThis.L = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops commented on docs/examples/accessibility/example.md:
```diff
@@ -3,16 +3,19 @@ layout: tutorial_frame
+	const marker = new Marker([50.4501, 30.5234], {alt: 'Kyiv'}).addTo(map)
 		.bindPopup('Kyiv, Ukraine is the birthplace of Leaflet!');
 
+	window.L = L; // only for debugging in the developer console
+	window.map = map; // only for debugging in the developer console
```
```suggestion
	globalThisL = L; // only for debugging in the developer console
	globalThis.map = map; // only for debugging in the developer console
```

jonkoops reviewed (commented):
Perhaps we can choose not to expose the globals in the examples? Is this required for any specific purpose except development debugging? If possible, I'd like to omit it.

Falke-Design commented on docs/_layouts/tutorial_frame.html:
```diff
@@ -10,7 +10,16 @@
+	<script type="importmap">
+		{
+			"imports": {
+				"leaflet": "https://unpkg.com/leaflet@{{ site.latest_leaflet_version}}/dist/leaflet.js"
```
Let us create a new issue for this an tackle it later. I don't have the experience to make sure that everything works. For example the generated hash is different for [cdnjs](https://cdnjs.com/libraries/leaflet). They are using `sha512` and we `sha256` https://leafletjs.com/download.html

jonkoops replied:
Fair enough, not going to hold this up for it.

Falke-Design commented on docs/docs/js/docs.js:
```diff
@@ -99,10 +99,18 @@ function clickOnAnchor(e) {
 	let scrollBy = elemHeader.nextSibling.offsetTop;
 
-	if (L.Browser.chrome && sameAnchor) {
+	if (chromeBrowser && sameAnchor) {
```
Let us tackle this in a separate issue too

jonkoops replied:
Sounds good, resolved.

jonkoops reviewed (approved):
Download page still needs a bit of rework, but not going to hold this up any longer than it needs to be, lgtm.

Falke-Design commented:
@jonkoops I updated the `download.md` file and added `importmap` and `modulepreload`. 
I'm honest, I needed a little bit of help from AI, because I never worked wirh modulepreload. It looks like, when I'm adding the `modulepreload` then the `importmap` is not working anymore. 

Can you please check the documentation and if the implementation of modulepreload is correct?

> Perhaps we can choose not to expose the globals in the examples? Is this required for any specific purpose except development debugging? If possible, I'd like to omit it.

I would like to keep the globals, because the tutorials are good bases to check something quick

Falke-Design commented:
@jonkoops some things that I noticed:

1. The Rollup bundle contains some `var`s
2. Without using the `L` namespace the user has some conflicts with the JS built in `Map` and `(Leaflet) Map` -> `new Map`. Maybe we should switch to a new name like `LeafletMap` (with an alias for `Map`)?

What do you think about that points?

jonkoops commented:
> The Rollup bundle contains some `var`s

This might be some compatibility stuff injected by Rollup, would you mind filing an issue for it? I think we can resolve this outside of the scope of this PR.

> Without using the `L` namespace the user has some conflicts with the JS built in `Map` and `(Leaflet) Map` -> `new Map`. Maybe we should switch to a new name like `LeafletMap` (with an alias for `Map`)?

This has been on my mind as well, and I was thinking we can keep `Map` around (perhaps deprecated), and have a new `LeafletMap` class around. This would be an odd one out compared to other classes, but I think that is fine.

Falke-Design commented:
> @jonkoops I updated the `download.md` file and added `importmap` and `modulepreload`. I'm honest, I needed a little bit of help from AI, because I never worked wirh modulepreload. It looks like, when I'm adding the `modulepreload` then the `importmap` is not working anymore.
> 
> Can you please check the documentation and if the implementation of modulepreload is correct?
> 
> > Perhaps we can choose not to expose the globals in the examples? Is this required for any specific purpose except development debugging? If possible, I'd like to omit it.
> 
> I would like to keep the globals, because the tutorials are good bases to check something quick

@jonkoops pls help ;)

jonkoops reviewed (approved):
The download page looks good to me now, we can always add further improvements later on.

Falke-Design commented:
> it looks like, when I'm adding the `modulepreload` then the `importmap` is not working anymore.

@jonkoops do you have an idea why this is happening?
